File: packages/gg-sandbox-python/src/gg/docs.py
"""Find what this session can call, and take a documentation view back out of the window.

The system prompt names modules and no function. `search` turns a keyword or a module id into
fully-qualified names, and a documentation view reads one of those names in full. Searching is always
bound; closing a documentation view is bought by a capability, and a run that did not enable it gets
`ApiErrorCode.UNAVAILABLE`.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from wit_world.imports import docs as wire

from ._registry import missing, operation
from .core import ApiError, ApiErrorCode, _call, _strings, _uint

__all__ = [
    "DocHit",
    "DocKind",
    "DocSearch",
    "close",
    "close_all",
    "search",
]


class DocKind(Enum):
    """Which of the three kinds of thing a documentation entry describes."""

    MODULE = "module"
    """A module a program imports, holding the functions it calls."""

    FUNCTION = "function"
    """A function a program calls."""

    TYPE = "type"
    """A type a function takes or hands back."""


@dataclass(frozen=True)
class DocHit:
    """One entry a search matched: enough to choose from, and no more."""

    key: str
    """The fully-qualified name a documentation view is opened by."""

    kind: DocKind
    """Whether it is a module, a function or a type."""

    module: str
    """The module it lives in: the one that publishes a function, or the one that declares a type.

    It is the module's own path, which is what `modules` filters on.
    """

    name: str
    """The name a program calls it by, or the type's own name."""

    summary: str
    """Its one-line brief, and only that. The rest is what a documentation view is for."""


@dataclass(frozen=True)
class DocSearch:
    """A page of search results, with the total behind it."""

    total: int
    """How many entries matched before paging, which tells a capped page from a complete answer."""

    offset: int
    """The offset this page starts at, echoed back."""

    hits: list[DocHit]
    """The page itself, best first."""


def _kind(value: str) -> DocKind:
    """The wire's word for a kind, as the enum a program compares against.

    gg owns both ends of this string and its set is closed at three, so a word that is not one of
    them is a mismatch between this SDK and the host rather than anything a program did — reported
    as such, rather than as the `ValueError` an enum lookup would otherwise raise out of the
    lowering.
    """
    try:
        return DocKind(value)
    except ValueError:
        raise ApiError(
            "search",
            ApiErrorCode.OTHER,
            f"gg reported a kind of documentation entry this SDK does not know: {value!r}",
        ) from None


@operation("docs.search")
def search(
    *,
    query: str | None = None,
    modules: list[str] | None = None,
    type: str | None = None,
    kind: DocKind | None = None,
    offset: int | None = None,
    limit: int | None = None,
) -> DocSearch:
    """Search every module, function and type this session can call, by keyword and by filter.

    Matching is a case-insensitive substring over names, signatures, briefs and detailed
    descriptions, so `docs` finds `open_docs_view`. Ranking is by the kind of evidence that matched —
    an entry whose own name matched outranks one that merely mentions the word in a paragraph — and a
    weaker kind never overtakes a stronger one however often it occurs.

    Only entries this run bound are returned. Every argument is optional and they compose: a query
    alone ranks the whole surface, a filter alone is a directory of what it names, and the two
    together search inside the filter.

    The page comes back as a value and is also opened as a view, under the selector `search results`.
    The next search replaces that view.

    Args:
        query: The words to match, as a case-insensitive substring. The default is no query at all,
            which is what turns a filter into a directory rather than a search.
        modules: The modules to look in, each named by its gg id — `files`, `views`, `docs` — or by
            the path a program imports it under — `gg.files` — and matched exactly. Several name a
            union: an entry in any one of them is a hit, so one call lists a whole granted surface.
            With no `query` it is those modules' whole directory. A name no module has matches
            nothing rather than failing, so an empty page means the filter found nothing, not that
            it was rejected. The default is no module filter at all.
        type: One type's name, narrowing to that type and to the functions that take or return it.
            Like `modules`, a name nothing declares matches nothing rather than failing.
        kind: Whether to return modules, functions or types. The default returns all three.
        offset: How many hits to skip, for reading past the first page. The default starts at the
            best hit.
        limit: The most hits to return. The default is gg's own page size and there is a ceiling
            above it. Zero is refused.

    Returns:
        The page that matched, best first. `total` counts every entry that matched before paging, so
            a page shorter than `total` is a page there is more of.

    Raises:
        ApiError: `invalid-argument` for a call with no query and no filter at all, and for a
            `limit` of zero. A `modules` entry or a `type` that names something gg does not hold
            matches nothing rather than failing.
    """
    found = _call(
        wire.search,
        query,
        _strings("search", "modules", modules),
        type,
        None if kind is None else kind.value,
        _uint("search", "offset", offset),
        _uint("search", "limit", limit),
    )
    return DocSearch(
        total=found.total,
        offset=found.offset,
        hits=[
            DocHit(
                key=hit.key,
                kind=_kind(hit.kind),
                module=hit.module,
                name=hit.name,
                summary=hit.summary,
            )
            for hit in found.hits
        ],
    )


@operation("docs.close")
def close(key: str) -> int:
    """Take one documentation view out of the context window, by the key it was opened under.

    The removal has no cascade: closing a type's view leaves every function view beside it, and
    closing a function's leaves its types. Nothing records why a view was opened, so a type that is
    closed is opened again by the next function that mentions it.

    Args:
        key: The fully-qualified name the view was opened under.

    Returns:
        How many views were taken away; a key that is not open closes `0` rather than failing.

    Raises:
        ApiError: `unavailable` under a run that did not enable closing documentation views.
    """
    return _call(wire.close_doc_view, key)


@operation("docs.close_all")
def close_all() -> int:
    """Take every documentation view out of the context window.

    The blanket form of `close`, on exactly the same terms and behind the same capability.

    Returns:
        How many documentation views went, and `0` rather than a failure when none was open.

    Raises:
        ApiError: `unavailable` under a run that did not enable closing documentation views.
    """
    return _call(wire.close_doc_views)


__getattr__ = missing(__name__, __all__)
"""What this module answers for a name it does not declare — see `gg._registry.missing`."""

File: packages/gg-sandbox-ruby/src/gg/docs.rb
# frozen_string_literal: true

module GG
  # Find what the agent can call, and take a documentation view back out of the window.
  #
  # The system prompt names modules and no function at all, so this is the module every other one is
  # reached through: `GG::Docs.search` turns a keyword or a module id into fully-qualified names, and
  # `GG::Views.open_docs_view` reads one of those names in full.
  #
  # Searching is bound in every program whatever a run enables, because an agent must always be able
  # to find the functions it does hold. Closing a documentation view is the exception and is bought
  # by a capability: opening one only ever appends to the prompt, while closing one rewrites its
  # middle, and those are different enough trades to be different decisions.
  module Docs
    extend Surface::Operations

    # Search every module, function and type the agent holds, by keyword and by filter.
    #
    # This is how a name is found. Matching is a case-insensitive substring over names, signatures,
    # briefs and detailed descriptions, so `docs` finds `open_docs_view`. Ranking is by the kind of
    # evidence that matched — an entry whose own name matched outranks one that merely mentions the
    # word in a paragraph — and a weaker kind never overtakes a stronger one however often it
    # occurs.
    #
    # Only entries this run bound are returned, so nothing a search finds is something the run
    # withheld. The filters compose with each other and with `query`.
    #
    # The page comes back as a value and is also opened as a view, under the selector
    # `search results`, so it can be read on the next turn without a program showing it to itself.
    # The next search replaces that view: it names what is being worked from rather than keeping a
    # record.
    #
    # @param query [String] The words to match, as a case-insensitive substring. It may be empty
    #   when at least one filter is given.
    # @param in_module [String, nil] One module's id — `files`, `views`, `docs` — matched exactly.
    #   An empty `query` with a module is that module's whole directory rather than a search. A
    #   module filter is a lookup, so a name no module has matches nothing rather than failing.
    # @param type [String, nil] One type's name, narrowing to that type and to the functions that
    #   take or return it. Like `in_module`, a name nothing declares matches nothing rather than
    #   failing.
    # @param kind [GG::Docs::DocKind, nil] `:module`, `:function` or `:type` — whether to return
    #   modules, functions or types. The default returns all three.
    # @param offset [Integer, nil] How many hits to skip, for reading past the first page. The
    #   default starts at the best hit.
    # @param limit [Integer, nil] The most hits to return. The default is gg's own page size and
    #   there is a ceiling above it, so comparing the hits against `total` is the only way to see a
    #   capped page. Zero is refused rather than read as "no cap".
    # @return [GG::Docs::DocSearch] the page that matched, best first, with the total behind it
    # @raise [GG::Core::ApiError] `:invalid_argument` for an empty query with no filter at all —
    #   nothing matched and nothing was asked for are different answers — for a `kind` that is none
    #   of `:module`, `:function` and `:type`, and for a `limit` of zero, which asks for a page that
    #   answers nothing. A module or type name gg does not hold is not among them: it matches
    #   nothing.
    def self.search(query, in_module: nil, type: nil, kind: nil, offset: nil, limit: nil)
      found = Wire.call("search", "docs", "search", [
                          query,
                          Wire.js(in_module),
                          Wire.js(type),
                          Wire.js(Wire.arm(Check.choice("search", "kind", kind,
                                                        [DocKind::MODULE, DocKind::FUNCTION,
                                                         DocKind::TYPE]))),
                          Wire.js(Check.uint("search", "offset", offset)),
                          Wire.js(Check.uint("search", "limit", limit))
                        ])
      DocSearch.new(
        total: Wire.field(found, "total"),
        offset: Wire.field(found, "offset"),
        hits: Wire.field(found, "hits").map { |hit| hit_of(hit) }
      )
    end
    operation :search, "docs.search"

    # Take one documentation view out of the context window, by the key it was opened under.
    #
    # The removal has no cascade: closing a type's view leaves every function view beside it, and
    # closing a function's leaves its types. Nothing records why a view was opened, so a type that
    # is closed is opened again by the next function that mentions it.
    #
    # @param key [String] The fully-qualified name the view was opened under.
    # @return [Integer] how many views were closed; a key that is not open closes `0` rather than
    #   failing
    # @raise [GG::Core::ApiError] `:unavailable` under a run that did not enable closing
    #   documentation views.
    def self.close(key)
      Wire.call("close", "docs", "closeDocView", [key])
    end
    operation :close, "docs.close"

    # Take every documentation view out of the context window and report how many went.
    #
    # The blanket form of `GG::Docs.close`, on exactly the same terms and behind the same
    # capability.
    #
    # @return [Integer] how many views were closed
    # @raise [GG::Core::ApiError] `:unavailable` under a run that did not enable closing
    #   documentation views.
    def self.close_all
      Wire.call("close_all", "docs", "closeDocViews", [])
    end
    operation :close_all, "docs.close_all"

    # One hit, as the class a program actually gets.
    #
    # @param hit [Object] the record the membrane returned
    # @return [GG::Docs::DocHit] the same hit, as the model-facing class
    # @api private
    def self.hit_of(hit)
      DocHit.new(
        key: Wire.field(hit, "key"),
        kind: Wire.symbol(Wire.field(hit, "kind")),
        in_module: Wire.field(hit, "module"),
        name: Wire.field(hit, "name"),
        summary: Wire.field(hit, "summary")
      )
    end
    private_class_method :hit_of

    # Which of the three kinds of thing a documentation entry describes.
    module DocKind
      # A module a program imports, with the functions it holds inside it.
      MODULE = :module

      # A function a program calls.
      FUNCTION = :function

      # A type a function takes or hands back.
      TYPE = :type
    end

    # One entry a search matched: enough to choose from, and no more.
    class DocHit
      include Value

      # @return [String] The fully-qualified name `GG::Views.open_docs_view` takes to read the whole
      #   entry.
      attr_reader :key

      # @return [GG::Docs::DocKind] Whether it is a `:module`, a `:function` or a `:type`.
      attr_reader :kind

      # The module it lives in.
      #
      # A module is its own, and a function has exactly one. A type shows every module in which a
      # function the agent can call mentions it — never one nothing is held in, and comma-separated
      # when there are several, which makes it a description rather than something to pass back as
      # `in_module`.
      #
      # @return [String] the module's id, or the ids of every module a type is mentioned in
      attr_reader :module

      # @return [String] The name a program calls it by, or the type's or the module's own name.
      attr_reader :name

      # @return [String] Its one-line brief, and only that. The rest is what a documentation view is
      #   for.
      attr_reader :summary

      # `in_module` rather than `module`, which Ruby reserves: the reader is `module` because that is
      # what the field is, and a keyword argument of that name is one no method body could read back.
      #
      # @api private
      def initialize(key:, kind:, in_module:, name:, summary:)
        @key = key
        @kind = kind
        @module = in_module
        @name = name
        @summary = summary
        freeze
      end
    end

    # A page of search results, with the total behind it.
    class DocSearch
      include Value

      # @return [Integer] How many entries matched before paging, which tells a capped page from a
      #   complete answer.
      attr_reader :total

      # @return [Integer] The offset this page starts at, echoed back.
      attr_reader :offset

      # @return [Array<GG::Docs::DocHit>] The page itself, best first.
      attr_reader :hits

      # @api private
      def initialize(total:, offset:, hits:)
        @total = total
        @offset = offset
        @hits = hits
        freeze
      end
    end
  end
end
